ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/deduplication/DeduplicationPeriodConverterSpec.scala

```diff
@@ -0,0 +1,168 @@
+import com.daml.ledger.api.domain.{ApplicationId, LedgerId, LedgerOffset}
+import com.daml.ledger.api.testing.utils.AkkaBeforeAndAfterAll
+import com.daml.ledger.api.v1.command_completion_service.{Checkpoint, CompletionStreamResponse}
+import com.daml.ledger.api.v1.ledger_offset
+import com.daml.ledger.participant.state.index.v2.IndexCompletionsService
+import com.daml.ledger.participant.state.kvutils.Conversions
+import com.daml.lf.data.Ref
+import com.google.protobuf.timestamp.Timestamp
+import org.mockito.MockitoSugar
+import org.scalatest.BeforeAndAfterEach
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.wordspec.AsyncWordSpec
+
+class DeduplicationPeriodConverterSpec
+    extends AsyncWordSpec
+    with Matchers
+    with MockitoSugar
+    with BeforeAndAfterEach
+    with AkkaBeforeAndAfterAll
+    with TestLoggers {
+  private val ledgerId = LedgerId("id")
+  private val indexCompletionsService: IndexCompletionsService = mock[IndexCompletionsService]
+  private val service = new DeduplicationPeriodConverter(ledgerId, indexCompletionsService)
+  private val offset = Ref.LedgerString.assertFromString("offset")
+  private val applicationId = ApplicationId(Ref.ApplicationId.assertFromString("id"))
+  private val parties = Set.empty[Ref.Party]
+  private val emptyResponse = CompletionStreamResponse()
+
+  override protected def afterEach(): Unit = reset(indexCompletionsService)
+
+  "return expected duration" in {
+    val submittedAt = Instant.now()
+    val response = Source.single(
+      emptyResponse.update(
+        _.checkpoint.update(
+          _.recordTime := Timestamp.fromJavaProto(
+            Conversions.buildTimestamp(submittedAt.minusSeconds(1))
+          ),
+          _.offset := ledger_offset.LedgerOffset(ledger_offset.LedgerOffset.Value.Absolute(offset)),
+        )
+      )
+    )
+    completionServiceReturnsResponse(response)
+    service
+      .convertOffsetToDuration(
+        offset,
+        applicationId,
+        parties,
+        submittedAt,
+      )
+      .map { result =>
+        result shouldBe Right(Duration.ofSeconds(1))
+      }
+  }
+
+  "return failure when there is an empty response" in {
+    completionServiceReturnsResponse(Source.empty)
+    service
+      .convertOffsetToDuration(
+        offset,
+        applicationId,
+        parties,
+        Instant.now(),
+      )
+      .map { result =>
+        result shouldBe Left(DeduplicationConversionFailure.CompletionAtOffsetNotFound)
+      }
+  }
+
+  "return failure when the checkpoint is missing" in {
+    completionServiceReturnsResponse(Source.single(emptyResponse))
+    service
+      .convertOffsetToDuration(
+        offset,
+        applicationId,
+        parties,
+        Instant.now(),
+      )
+      .map { result =>
+        result shouldBe Left(DeduplicationConversionFailure.CompletionCheckpointNotAvailable)
+      }
+  }
+
+  "return failure when the checkpoint misses the record time" in {
```

Comment:
Nit: `checkpoint misses the record time` => `checkpoint is missing a record time` or `checkpoint does not have an associated record time`